__tests__/index.ts

```diff
@@ -80,6 +80,106 @@ describe('test suites', () => {
   );
 });
 
+describe('issue 58 - key and value being identical', () => {
+  it('should handle duplicate but equal `Map` entries', () => {
+    const mapA = new Map<any, any>([
+      [{ b: 'c' }, 2],
+      [{ b: 'c' }, 2],
+    ]);
+    const mapB = new Map<any, any>([
+      [{ b: 'c' }, 2],
+      [{ b: 'c' }, 2],
+    ]);
+
+    const result = deepEqual(mapA, mapB);
+
+    expect(result).toBe(true);
+  });
+
+  it('should handle the first `Map` entry being unequal', () => {
+    const mapA = new Map<any, any>([
+      [{ b: 'c' }, 2],
+      [{ b: 'c' }, 2],
+    ]);
+    const mapB = new Map<any, any>([
+      [{ b: 'c' }, 2],
+      ['foo', 'different'],
+    ]);
+
+    const result = deepEqual(mapA, mapB);
+
+    expect(result).toBe(false);
+  });
+
+  it('should handle the last `Map` entry being unequal', () => {
+    const mapA = new Map<any, any>([
+      [{ b: 'c' }, 2],
+      [{ b: 'c' }, 2],
+    ]);
+    const mapB = new Map<any, any>([
+      ['foo', 'different'],
+      [{ b: 'c' }, 2],
+    ]);
+
+    const result = deepEqual(mapA, mapB);
+
+    expect(result).toBe(false);
+  });
+
+  it('should handle an intermediary `Map` entry being unequal', () => {
+    const mapA = new Map<any, any>([
+      ['foo', 'different'],
+      [{ b: 'c' }, 2],
+      [{ b: 'c' }, 2],
+    ]);
+    const mapB = new Map<any, any>([
+      ['foo', 'different'],
+      ['foo', 'different'],
+      [{ b: 'c' }, 2],
+    ]);
+
+    const result = deepEqual(mapA, mapB);
+
+    expect(result).toBe(false);
+  });
+
+  it('should handle duplicate but equal `Set` entries', () => {
+    const setA = new Set<any>([{ b: 'c' }, { b: 'c' }]);
+    const setB = new Set<any>([{ b: 'c' }, { b: 'c' }]);
+
+    const result = deepEqual(setA, setB);
+
+    expect(result).toBe(true);
+  });
+
+  it('should handle the first `Set` entry being unequal', () => {
+    const setA = new Set<any>([{ b: 'c' }, { b: 'c' }]);
+    const setB = new Set<any>([{ b: 'c' }, 'foo']);
+
+    const result = deepEqual(setA, setB);
```

Comment:
For "first", "last" and "intermediary"  test we could create "reversed" tests with `setA` and `setB` switched in the call of the function. 
Since we have two loops in `areSetsEqual`, the order is important. For example, without fixing `areSetsEqual` and just changing the order in which we iterate (switching a and b) the tests will pass.